#182210052 make trip request

K-Ds opened the pull request:
#### What does this PR do?
- Add Trip Request Feature

#### Description of Task to be completed?
- A requester(user) should be able create a trip request
- A requester should be able to view all of his/her trip requests
- A manager should be able to view all trip requests of all his Direct employee
- A requester should be able to delete or update  trip request if it still has a pending status
    
#### How should this be manually tested?
- Make sure there are accommodations available in database
- Using swagger docs(web) to login and create trip requests

#### Any background context you want to provide?
none

#### What are the relevant pivotal tracker stories?
182210052

Ihimbazwe committed:
Added productuin mode to heroku

resolve conflicts

Ihimbazwe committed:
Added production mode to heroku

Ihimbazwe committed:
Added production mode to heroku

Ihimbazwe committed:
Added production mode to heroku

Ihimbazwe committed:
Added production mode to heroku

Ihimbazwe committed:
Added production mode to heroku

Ihimbazwe committed:
Added production mode to heroku

Ihimbazwe committed:
Added  Fixed babel config  dependencies  to heroku

resolve conflicts

munezerobagira committed:
Fix: Hound CI review error

-remove the parser

K-Ds committed:
Models and routes

K-Ds committed:
Implementation of validations

Ihimbazwe committed:
fix(Conflicting files): removed the cors in index and packageloc

resolve conflicts

K-Ds committed:
#####feat(trip request): CRUD for trip request

- implement creating a trip request
- implemet get all trip requested owned by user
- implement editing & deleting trip request if status is pending

pacomesimon committed:
test(helper): test the verification email function

Ihimbazwe committed:
fix(missing): Added the Procfile for which was missing in this feat

pacomesimon committed:
test(helper): change the env variables

pacomesimon committed:
chore(helper): clean the codes

Ihimbazwe committed:
chore: modified the start script and procfile

resolve conflicts

nyakamwe committed:
fix(coverage): use env variables in tests

Himbaza123h committed:
feat(Authentication): correction based on PR's comments

pacomesimon committed:
chore(helper): enhance DRY and reusability

pacomesimon committed:
chore(helper): enhance DRY and reusability

pacomesimon committed:
chore(helper): update readme

nyakamwe committed:
fix(coverage): Set decrease for failure

nyakamwe committed:
fix(coverage): Set decrease for failure

Ihimbazwe committed:
chore: modified start script in package json for heroku deployment

resolve conflicts

nyakamwe committed:
fix(coverage): import dotenv in helper

nyakamwe committed:
fix(coverage): describe get default helper

nyakamwe committed:
fix(coverage): use cross-env for multi db

nyakamwe committed:
fix(coverage): use cross-env for multi db

nyakamwe committed:
fix(coverage): add test to coverage

Himbaza123h committed:
feat(Authentication): resolving conflicts

K-Ds committed:
Test(tripRequest): Creating Tests

Creating tests for the trip requests and correcting code where due

#182210052

Himbaza123h committed:
feat(Auth check middleware):

munezerobagira committed:
feat(logout) logout functionality

- fix test

munezerobagira committed:
feat(logout) logout functionality

- fix test

munezerobagira committed:
feat(logout) logout functionality

- fix filename

K-Ds committed:
Feat(Trp Request): Add User model

- Integrate User model from develop branch
- Change terms and names in code to match new User model

K-Ds committed:
feature(helpers):add verificationEmail function

K-Ds committed:
fix(coverage): Add files to be included or excluded

- adding all js file found in src to be included
- exclude files from test folder

[coveralls #182209989]

K-Ds committed:
fix(coverage): Do test for boost up coverage

- Test database connection for testing
- writing test for root and users routes

[Coverall #182209989]

K-Ds committed:
fix(coverage): Apply object destructuring for coverage boost

K-Ds committed:
fix(coverage): Add before hook

- Adding connect to database in before hook

K-Ds committed:
Testing heroku version 1

resolve conflicts

K-Ds committed:
Added production mode to heroku

resolve conflicts

K-Ds committed:
Added production mode to heroku

resolve conflicts

K-Ds committed:
fix(coverage): boost coverage

K-Ds committed:
fix(coverage): fix url for database

K-Ds committed:
fix(coverage): import dotenv in helper

K-Ds committed:
fix(coverage): use cross-env for multi db

K-Ds committed:
Testing heroku version 1

K-Ds committed:
Added productuin mode to heroku

K-Ds committed:
Added production mode to heroku

K-Ds committed:
Added production mode to heroku

K-Ds committed:
Added  removed runtime  to heroku

K-Ds committed:
Added  removed unnecessary dependencies  to heroku

K-Ds committed:
Added  Fixed babel config  dependencies  to heroku

K-Ds committed:
fix(Conflicting files): removed the cors in index and packageloc

K-Ds committed:
fix(missing): Added the Procfile for which was missing in this feat

K-Ds committed:
fix(coverage): add test to coverage

K-Ds committed:
ft(user-register):user registration

- create user registration route
- test the endpoint
- document the route

[Delivers #1822100043]

K-Ds committed:
feature(auth): enables the email verification route

[Finishes #182210044]

K-Ds committed:
feat(login): implement login

- added test for login

K-Ds committed:
feat(login):fix login

- fixed test for login

K-Ds committed:
feat(login):fix login

- fixed test for login

K-Ds committed:
refactor(auth): implements email verification

 - integrates email verification with signin and signup tests

[Finishes #182210044]

K-Ds committed:
ft: logout functionality

- add the user session model

K-Ds committed:
feat(logout) implement logout functionality

- updated session model

added the test

K-Ds committed:
ft: logout functionality

- add the user session model

K-Ds committed:
feat(logout) implement logout functionality

- updated session model

added the test

K-Ds committed:
feat(logout) logout functionality

- add script to run undo migration after test

add test for logout

K-Ds committed:
feat(logout) logout functionality

- added the routes for sessions

K-Ds committed:
change-scripts

K-Ds committed:
feat(Authentication): Generating JWT for user

K-Ds committed:
feat(Authentication): resolving conflicts

K-Ds committed:
feat(Authentication): resolving conflicts

K-Ds committed:
feat(Auth check middlewares): Coveralls test

K-Ds committed:
feat(Auth check middleware): fixing errors with circle ci checks

K-Ds committed:
feat(logout) logout functionality

- added more test

K-Ds committed:
feat(logout) logout functionality

- debug the test online

K-Ds committed:
documentation(swagger): add Swagger docs for acc signup and verification

K-Ds committed:
test(emailing): allows usage of invalid sendgrid key for tests

K-Ds committed:
feat(logout) logout functionality

- change from x-auth-token to authorization

K-Ds committed:
Models and routes

K-Ds committed:
Test(tripRequest): Creating Tests

Creating tests for the trip requests and correcting code where due

K-Ds committed:
fix(TripRequest): Integrate User

- Change trip request code to match User in develop

K-Ds committed:
<test>(Trip Request): Add Tests

- Add tests to trip request routes, controlles and services

-Add Integrate test to other feature where change is needed

[Finishes #182210052]

K-Ds committed:
Merge branch 'develop' into ft-make-trip-request-182210052

K-Ds committed:
fix(Trip Request): Add trip Request route

- Add trip reqeust routes definition in routes/index.js

K-Ds committed:
Merge branch 'develop' into ft-make-trip-request-182210052

K-Ds committed:
fix(Trip Request): Fix packages

- Fix bug in package.json
- Fix bug in package-lock.json

K-Ds committed:
test(Trip Request): Add tests

- Add tests to trip request routes and fix failed tests

K-Ds committed:
Docs(Trip Request): Add documentation

- Add swagger documentation to for trips

nyakamwe commented on src/services/trip.services.js:
```diff
@@ -0,0 +1,115 @@
+export const getAllTripRequests = async (user) => {
+  let result;
+
+  if (user.user_role == "6927442b-84fb-4fc3-b799-11449fa62f52") {
+    result = await TripRequest.findAll({
```
Instead of hardcoding this id, you can use findRoleById(roleId) service in roleService.js for all hardcode ids in this tip.service file

K-Ds replied:
Ok will do

nyakamwe commented on src/database/migrations/dummy-accommodations.js:
```diff
@@ -0,0 +1,123 @@
+    });
+
+    await queryInterface.createTable("trip_requests", {
+      id: {
+        allowNull: false,
+        primaryKey: true,
+        type: Sequelize.INTEGER,
+        autoIncrement: true,
```
TTL suggested me to use UUID instead of INTEGER, so I would like to recommend too

nyakamwe reviewed (commented):
left some comments in service and migrations

habaumugisha1 commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "scripts": {
     "build": "npx babel src --out-dir lib",
-    "test": "cross-env NODE_ENV=test npm run unmigrate && cross-env NODE_ENV=test npm run migrate && cross-env NODE_ENV=test nyc --reporter=text --reporter=html  mocha --recursive --require @babel/polyfill --require @babel/register ./src/test/*.js --timeout 100000 --exit ",
+    "test": "cross-env NODE_ENV=test npm run unmigrate && cross-env NODE_ENV=test npm run migrate && cross-env NODE_ENV=test npm run seed-roles && cross-env NODE_ENV=test nyc --reporter=text --reporter=html  mocha --recursive --require @babel/polyfill --require @babel/register ./src/test/*.js --timeout 100000 --exit ",
```
have you made a rebase? @K-Ds

K-Ds committed:
fix(trip Request) : change according to review

- Fix Date of return allow null
- Rename trip routes
- Add Associations

K-Ds committed:
feat(location) : Add Location Entity

- Create models and migrations for location
- Create routes and services for Location
- Create docs and tests for location

K-Ds committed:
fix(package.json): fix test script

-Add unmigrate on end of script

K-Ds committed:
fix(package.json): fix test script

- Add unmigrate to the beginnig of test script

K-Ds committed:
fix(package-lock): fix conflicts

- Resolve conficts btwn develop and this branch

K-Ds committed:
Merge branch 'develop' of https://github.com/atlp-rwanda/winners-c8-bn-be into ft-make-trip-request-182210052

K-Ds committed:
docs(trip request) : Change some names

- Change departure to departureId
- Change destination to destinationId

habaumugisha1 commented on src/controllers/locationControllers.js:
```diff
@@ -0,0 +1,87 @@
+    const locationId = req.params.id;
+    const location = await locationServices.deleteLocation(locationId);
+    return successResponse(res, 201, "Successfully deleted location");
+  } catch (err) {
```
what will happen if a location ID is invalid?

K-Ds replied:
As in the lines below, if deleteLocation service finds that the location id is invalid it throws an error which is caught by the catch block below

habaumugisha1 commented on src/controllers/tripRequests.js:
```diff
@@ -0,0 +1,164 @@
+
+  const result = await tripServices.getAllTripRequests(user);
+  // if (result.length === 0) {
+  //   return res.status(200).json({ message: "No Trip Requests" });
```
Remove commented codes
